### ISSUE TITLE
Inconsistent behavior with a custom query utilizing WHERE IN

### ISSUE THREAD

**Issue by knyghtryda** [simolus3/moor#236]
I have this query declared inside my @UseDao:
```
'filteredContacts': 'SELECT contacts.* FROM contacts '
      'INNER JOIN contacts_tags_map ON contacts.id = contacts_tags_map.contact_id '
      'LEFT JOIN tags ON tags.id = contacts_tags_map.tag_id '
      'WHERE contacts.pending = 0 AND tags.id IN ? '
      'GROUP BY contacts.id HAVING COUNT(tags.id) = ? '
      'ORDER BY ?'
```

This is just a variation on a filtering query:
```sql
SELECT a.* FROM ObjectA a 
INNER JOIN ObjectA_ObjectB ab ON a.id = ab.objectA_id 
JOIN ObjectB b ON b.Id = ab.objectB_id  
WHERE b.id IN (1,2,3) 
GROUP BY a.id HAVING COUNT(b.id) = 3;
```

However, this query is behaving very inconsistently between builds and hot reloads.  Most of the time this query returns nothing, even if the inputs I know should return some values.  If I hardcode some of the parameters (like the IN, or the COUNT(...) ) then this query returns the correct value, but once I revert this to accepting parameters it fails again.  Here's the generated code for reference.

```dart
Selectable<Contact> filteredContactsQuery(
      List<int> var1, int var2, String var3) {
    var $arrayStartIndex = 1;
    final expandedvar1 = $expandVar($arrayStartIndex, var1.length);
    $arrayStartIndex += var1.length;
    return customSelectQuery(
        'SELECT contacts.* FROM contacts INNER JOIN contacts_tags_map ON contacts.id = contacts_tags_map.contact_id LEFT JOIN tags ON tags.id = contacts_tags_map.tag_id WHERE contacts.pending = 0 AND tags.id IN ($expandedvar1) GROUP BY contacts.id HAVING COUNT(tags.id) = ? ORDER BY ?',
        variables: [
          for (var $ in var1) Variable.withInt($),
          Variable.withInt(var2),
          Variable.withString(var3)
        ],
        readsFrom: {
          contacts,
          contactsTagsMap,
          tags
        }).map(_rowToContact);
  }
```

Any insight into what might be happening?  I have tested this query on sqlfiddle and it seems to be working correctly.  Also, I'm not quite ready to move everything to .moor yet, as I have many table definitions in dart, but if there's a bug with how custom queries are being parsed or interpreted then I will bite the bullet and migrate.  I tried to just migrate the query by itself into a .moor but it looks like .moors get processed before dart files so its complaining that the tables generated by my dart file don't exist yet for the query to use.

**Comment by simolus3** [comment]
We've had some issues on those `x IN ?` queries, see #156. The error was caused by the sqflite library which `moor_flutter` wraps. That should have been fixed in version 1.1.6+5 of sqflite though. Can you check the sqflite version you're using? The exact version should be listed in `pubspec.lock`.

> Also, I'm not quite ready to move everything to .moor yet, as I have many table definitions in dart

That's probably a good choice for now - as you've noticed, we can't reliably read Dart files while processing moor files yet. This is a limitation with Dart's build system. I've submitted a PR that could fix this, so I hope this gets better eventually.
Using moor files would probably not have an effect on this issue in particular, it would generate the same or very similar code.

**Comment by knyghtryda** [comment]
My sqflite version is 1.1.7+2
```
sqflite:
    dependency: "direct main"
    description:
      name: sqflite
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.7+2"
```

**Comment by knyghtryda** [comment]
Tried with 1.1.6+5 just to make sure it wasn't a regression and the results are the same, query still doesn't work.  I'm currently trying to work around this with just a generated OR statement of all the elements that was being passed to IN but that doesn't seem to work for custom queries as the builder misclassifies my parameter type as a boolean instead of a string.  `AND ?` always gets classified as a boolean, and if I just throw in a `?` randomly the builder errors out.  Any ideas?

**Comment by Mike278** [comment]
You cannot bind column names or any SQL keywords with `ORDER BY ?` - only values. `ORDER BY` works with both column names and indexes though, so maybe SQLite is trying its best to interpret the `String var3` as a column index which produces the inconsistent results you're seeing.

**Comment by knyghtryda** [comment]
I've tried removing the ORDER BY and that had no effect on the query.  I'm not quite sure what you mean about the ORDER BY though.  What get's passed in is a string and you end up with `ORDER BY column_name ASC`.  What would be a valid input if its not a column name?

**Comment by simolus3** [comment]
Sorry, I didn't see the `ORDER BY ?` at first.

> and if I just throw in a `?` randomly the builder errors out

You can view the detailed stacktrace with `flutter pub run build_runner build -v`. Randomly adding a variable is likely a syntax error, so I wouldn't be surprised about the parser complaining.

Perhaps there's a misunderstanding on how `?` works in sqlite and moor. When sqlite parses a query with `?` in them, it sees them as an [expression](https://www.sqlite.org/lang_expr.html). We can set the value of that expression via [functions](https://www.sqlite.org/c3ref/bind_blob.html), which is what moor does with variables. Note that neither columns or table names are expressions, so variables can't be used for them.

>  What get's passed in is a string and you end up with `ORDER BY column_name ASC`. What would be a valid input if its not a column name?

The question mark gets bound to a variable, so the actual ordering clause is `ORDER BY 'column_name ASC'`. That string is the same for all rows of course, so all rows are considered equal by that ordering clause. [sqlite docs](https://www.sqlite.org/lang_select.html#orderby) say that "The order in which two rows for which all ORDER BY expressions evaluate to equal values are returned is undefined". So that could explain the inconsistencies you're seeing with that query.

------

We've added a new feature that let's you inline moor's Dart api into compiled statements (see [here](https://moor.simonbinder.eu/docs/using-sql/moor_files/#dart-components-in-sql)). That was designed to be used with moor files, but I think it should work for annotation-defined queries as well. If you write something like
```dart
'filteredContacts': 'SELECT contacts.* FROM contacts '
      'INNER JOIN contacts_tags_map ON contacts.id = contacts_tags_map.contact_id '
      'LEFT JOIN tags ON tags.id = contacts_tags_map.tag_id '
      'WHERE contacts.pending = 0 AND tags.id IN ? '
      'GROUP BY contacts.id HAVING COUNT(tags.id) = ? '
      r'ORDER BY $order'
```
Then moor should generate a `filteredContactsQuery(List<int> var1, int var2, List<OrderingTerm> order)` method. You could then pass an `[OrderingTerm(expression: CustomExpression('contacts.my_column_name'), mode: OrderingMode.asc)]` for that list. The difference is that a `CustomExpression` is written into the query directly (so you have to take care of sql injections and whatnot), but that let's you use insert any sql into the query.

Similarly, using `AND ?` gets recognized as a boolean (because it's a variable that should have a boolean value), but `AND $predicate` gets recognized as a `Expression<bool, BoolType>`, for which you can use the `and`, `or` and `not` functions.

**Comment by Mike278** [comment]
I think I'm actually running into this issue too. I'll try and come up with a better repro later, but my query looks something like this:
```sql
SELECT * FROM my_table
WHERE
  (CASE WHEN '-1' IN :some_col THEN 1 ELSE some_col IN :some_col END) AND
  (CASE WHEN -1 IN :another_col THEN 1 ELSE another_col IN :another_col END)
LIMIT :limit_
OFFSET :offset_;
```
and gets expanded at runtime like:
```sql
SELECT * FROM my_table
WHERE
  (CASE WHEN '-1' IN (?1) THEN 1 ELSE some_col IN (?1) END) AND
  (CASE WHEN -1 IN (?2) THEN 1 ELSE another_col IN (?2) END)
LIMIT :limit_
OFFSET :offset_ 

...with args [-1, -1, -1, -1]
```

My unit test that uses the `VmDatabase` executor returns rows as expected, but copying that unit test into a flutter widget (therefore using the `FlutterQueryExecutor`) does not return any rows. I haven't completely ruled out any other app-level factors, but it seems to be a similar issue to #156

**Comment by simolus3** [comment]
Thanks, I'm able to reproduce this with a very minimal example.
This ffi code finds a row:
```dart
import 'package:moor_ffi/database.dart';

void main() {
  final db = Database.memory();

  db.execute('CREATE TABLE repro (a TEXT, b INT);');
  db.prepare('INSERT INTO repro VALUES (?, ?)')
    ..execute(['text', 42])
    ..close();

  final select = db.prepare('''
SELECT * FROM repro
WHERE
  (CASE WHEN '-1' IN (?1) THEN 1 ELSE a IN (?1) END) AND
  (CASE WHEN -1 IN (?2) THEN 1 ELSE b IN (?2) END)
   ''');
  final result = select.select(['-1', -1]);
  print('got ${result.length} rows!'); // 1
}
```
Whereas this sqflite code does not:
```dart
import 'package:sqflite/sqflite.dart';

void repro() async {
  final db = await openDatabase(':memory:');

  await db.execute('CREATE TABLE repro (a TEXT, b INT);');
  await db.execute('INSERT INTO repro VALUES (?, ?)', ['text', 42]);

  final result = await db.rawQuery('''
SELECT * FROM repro
WHERE
  (CASE WHEN '-1' IN (?1) THEN 1 ELSE a IN (?1) END) AND
  (CASE WHEN -1 IN (?2) THEN 1 ELSE b IN (?2) END)
   ''', [
    '-1',
    -1,
  ]);

  print('got ${result.length} rows!'); // 0
}
```

I'm pretty sure I know what's causing this, but it's very hard to blame sqflite for that. On Android, all parameters must be provided as a string (see [this method](https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase.html#rawQuery(java.lang.String,%20java.lang.String[]))). So a `-1` would actually be bound as a `"-1"`, and `-1 IN ('-1')` is false. It looks like sqldelight uses `androidx.sqlite`, perhaps for this reason.

**Comment by Mike278** [comment]
> On Android, all parameters must be provided as a string

!!!
I'm surprised this issue [isn't](https://stackoverflow.com/questions/14940637/how-to-search-a-database-with-integer-types-in-android) [more](https://stackoverflow.com/questions/18746149/android-sqlite-selection-args-with-int-values) [common](https://stackoverflow.com/questions/18013912/selectionargs-in-sqlitequerybuilder-doesnt-work-with-integer-values-in-columns). Can't wait for dart:ffi to be stable.

@knyghtryda I'm guessing it's the `COUNT(tags.id) = ?` part of your query that's suffering from this issue. Maybe you could try `COUNT(tags.id) = CAST(? as INTEGER)`.

**Comment by simolus3** [comment]
The cast is a good idea, but unfortunately the generator doesn't handle casts yet. I'll expand the documentation on how to migrate to `moor_ffi`, the current state with integers on Android is pure madness.

Some background: Sqflite tries to work around that "parameters are strings"-limitation by replacing the variable with the actual value before executing the statement. So for `runSelect("SELECT ... WHERE x = ?", [3])`, sqflite would actually run `SELECT ... WHERE x = 3` without  any args. Sqflite doesn't recognize args with explicit indexes (which is why `runSelect('SELECT ?1', [1])` used to return 11), so now it stops inlining whenever after it sees an explicit variable. There's probably a similar bug when mixing named and unnamed args.

> Can't wait for dart:ffi to be stable.

`dart:ffi` is included in the public docs for Dart 2.6, it was hidden in 2.5. I guess that's good news for api stability. I think I'd actually recommend `moor_ffi` over `moor_flutter` for new users, precisely because we don't have to deal with those Android apis. The only downsides I see are long first build times and slightly increased app sizes (fairly minimal when using bundles or splitting per abi). But having the same sqlite version with the same compilation options on all apps, together with a sane api is just :pray:

**Comment by Mike278** [comment]
iirc last time I tried to switch to ffi I was blocked by https://github.com/flutter/flutter/issues/39916, but looks like that’s fixed really soon! 
Definitey looking forward to more direct access to the sqlite api 👍

**Comment by knyghtryda** [comment]
@simolus3 I don't think compiled statements is working correctly for queries generated from a DAO.  I'm trying to generate `'allContactsOrder': r'SELECT * FROM contacts ORDER BY $order',` as a test and it's giving me a generator parse error 
`Error while trying to parse allContactsOrder: line 1, column 39: Error: Could not parse this expression}
  ╷
1 │ SELECT * FROM contacts ORDER BY $order
  │                                       ^
`

**Comment by simolus3** [comment]
The parsing error when using `$order` in a `UseMoor` annotation is fixed in `moor_generator` version 2.2.0.

The `WHERE foo IN ?` madness is still present in `moor_flutter`, there's not much we can do about that unfortunately. However, it looks like the ffi issues with Flutter are fixed, and switching to `moor_ffi` would fix the problem.
